src/mame/brother/ax145.cpp

```diff
@@ -0,0 +1,274 @@
+	// devices
+	required_device<hd64180rp_device> maincpu;
+	required_device<hd44780_device> lcdc;
+	memory_share_creator<uint8_t> ram;
+	required_region_ptr<uint8_t> rom;
+
+	// valid values (bei IO3000=0x0b,0x07) (read_config @ 0x14c87): 0 = german => 0, 1 = german => 1, 2 = espanol => 2, 4 (gehäusedeckel offen) => 3, 8 = francais => 4, 16 = german => 5
+	static constexpr uint8_t id = 1;
+
+	// config switch
+	uint8_t io_3000{};
+
+	// bit 2: R/~W (Read / Not Write)
+	// bit 1: RS (Register Select)
+	// bit 0: E (Enable)
+	uint8_t lcd_signal{};
+
+	// driver_device overrides
+	void machine_start() override;
+	void machine_reset() override;
+	void video_start() override;
+
+	void map_program(address_map& map) {
+		map(0x00000, 0x01fff).rom();
+		map(0x04000, 0x1ffff).rom();
+		map(0x40000, 0x5ffff).bankr("dictionary");
+		// RAM is installed in machine_start()
+	}
+
+	void map_io(address_map& map) {
+		map.global_mask(0xffff);
+		map(0x0000, 0x003f).noprw(); // Z180 internal registers
+		//map(0x0040, 0x00ff).rw(FUNC(ax145_state::illegal_io_r), FUNC(ax145_state::illegal_io_w));
+		//map(0x2000, 0x2000).w(TODO);
+		//map(0x2800, 0x2800).w(TODO);
+		map(0x3000, 0x3000).w(FUNC(ax145_state::io_3000_w));
+		map(0x3800, 0x3800).r(FUNC(ax145_state::io_3800_r)); // config lower 4 bits
+		map(0x4000, 0x4000).r(FUNC(ax145_state::io_4000_r)); // config upper 4 bits
+		map(0x5000, 0x5000).w(FUNC(ax145_state::dictionary_bank_w));
+		map(0x5800, 0x5800).w(FUNC(ax145_state::lcd_signal_w));
+		map(0x6000, 0x6000).rw(FUNC(ax145_state::lcd_data_r), FUNC(ax145_state::lcd_data_w));
+		//map(0x6800, 0x6800).w(TODO);
+		//map(0x7000, 0x7000).w(TODO);
+		//map(0x7800, 0x7800).w(TODO);
+		//map(0x8000, 0x8000).w(TODO);
+		//map(0x8800, 0x8800).w(TODO);
+		//map(0x9000, 0x9000).r(TODO);
+		map(0xb000, 0xb000).r(FUNC(ax145_state::irq_ack_r));
+		//map(0xb800, 0xb800).rw(TODO, TODO);
+	}
+
+	void palette(palette_device &palette) const {
+		palette.set_pen_color(0, rgb_t(138, 146, 148));
+		palette.set_pen_color(1, rgb_t(92, 83, 88));
+	}
+
+	// IO
+	void io_3000_w(uint8_t data) { io_3000 = data; }
+
+	// should probably return something different depending on io_3000
+	// 0x0014a9 also reads but io_3000=0xfe,0xff...
+	uint8_t io_3800_r() {
+		//space.device().logerror("%s: IO 3000=%02x\n", pc(), io_3000);
+
+		// config lower 4 bits
+		return (~id) & 0x0f;
+	}
+
+	uint8_t io_4000_r() {
+		//space.device().logerror("%s: IO 3000=%02x\n", pc(), io_3000);
+
+		// config upper 4 bits
+		return (~id) >> 4;
+	}
+
+	void lcd_signal_w(uint8_t data) {
+		lcd_signal = data;
+	}
+
+	uint8_t lcd_data_r() {
+		if(BIT(lcd_signal, 1)) // RS
+			return lcdc->data_r();
+		else
+			return lcdc->control_r();
+	}
+	void lcd_data_w(uint8_t data) {
+		if(BIT(lcd_signal, 0)) { // E
+			if(BIT(lcd_signal, 1)) // RS
+				lcdc->data_w(data << 4);
+			else
+				lcdc->control_w(data << 4);
+		}
+	}
+
+	void dictionary_bank_w(uint8_t data) {
+		membank("dictionary")->set_entry(data & 0x03);
+	}
```

Comment:
This is a tag map lookup on the hot path.  Please use an object finder to find the memory bank.